### ISSUE TITLE
packer/communicator/ssh.(*comm).Download panic: not implemented yet

### ISSUE THREAD

**Issue by dkoudela** [mitchellh/packer#1903]
I have implemented a custom provisioner for downloading files from the running image before it is terminated.

Unfortunately, part of the communicator is not implemented yet.

Would it be possible to implement the functionality soon?

```
2015/02/06 16:02:11 ui: ==> vmware: Downloading /var/log/messages => vm-messages
2015/02/06 16:02:11 packer-provisioner-filedownload.exe: 2015/02/06 16:02:11 Opening the target file: vm-messages
2015/02/06 16:02:11 [INFO] 0 bytes written for 'downloadWriter'
2015/02/06 16:02:11 packer-builder-vmware-iso.exe: panic: not implemented yet
2015/02/06 16:02:11 packer-builder-vmware-iso.exe: 
2015/02/06 16:02:11 packer-builder-vmware-iso.exe: goroutine 88 [running]:
2015/02/06 16:02:11 packer-builder-vmware-iso.exe: runtime.panic(0x7a6c20, 0xc082071e10)
2015/02/06 16:02:11 packer-builder-vmware-iso.exe: 	/Users/mitchellh/code/3rdparty/go/src/pkg/runtime/panic.c:279 +0x11f
2015/02/06 16:02:11 packer-builder-vmware-iso.exe: github.com/mitchellh/packer/communicator/ssh.(*comm).Download(0xc082171d40, 0xc08200b2a0, 0x11, 0x135d58, 0xc082124140, 0x0, 0x0)
2015/02/06 16:02:11 packer-builder-vmware-iso.exe: 	/Users/mitchellh/code/go/src/github.com/mitchellh/packer/communicator/ssh/communicator.go:174 +0x7f
2015/02/06 16:02:11 packer-builder-vmware-iso.exe: github.com/mitchellh/packer/packer/rpc.(*CommunicatorServer).Download(0xc08218e4c0, 0xc08200b1a0, 0xc082085770, 0x0, 0x0)
2015/02/06 16:02:11 packer-builder-vmware-iso.exe: 	/Users/mitchellh/code/go/src/github.com/mitchellh/packer/packer/rpc/communicator.go:252 +0x14e
2015/02/06 16:02:11 packer-builder-vmware-iso.exe: reflect.Value.call(0x82bc20, 0x89ff58, 0x0, 0x130, 0x8f5ed0, 0x4, 0xc08201d9e0, 0x3, 0x3, 0x0, ...)
2015/02/06 16:02:11 packer-builder-vmware-iso.exe: 	/Users/mitchellh/code/3rdparty/go/src/pkg/reflect/value.go:563 +0x1217
2015/02/06 16:02:11 packer-builder-vmware-iso.exe: reflect.Value.Call(0x82bc20, 0x89ff58, 0x0, 0x130, 0xc08201d9e0, 0x3, 0x3, 0x0, 0x0, 0x0)
2015/02/06 16:02:11 packer-builder-vmware-iso.exe: 	/Users/mitchellh/code/3rdparty/go/src/pkg/reflect/value.go:411 +0xde
2015/02/06 16:02:11 packer-builder-vmware-iso.exe: net/rpc.(*service).call(0xc0820a4320, 0xc0820fc400, 0xc082071cb8, 0xc0820ce000, 0xc08218e620, 0x782680, 0xc08200b1a0, 0x0, 0x160, 0x782d40, ...)
2015/02/06 16:02:11 packer-builder-vmware-iso.exe: 	/Users/mitchellh/code/3rdparty/go/src/pkg/net/rpc/server.go:382 +0x1ce
2015/02/06 16:02:11 packer-builder-vmware-iso.exe: created by net/rpc.(*Server).ServeCodec
2015/02/06 16:02:11 packer-builder-vmware-iso.exe: 	/Users/mitchellh/code/3rdparty/go/src/pkg/net/rpc/server.go:453 +0x42e
```

**Comment by vtolstov** [comment]
I'm try to implement this,but can't find complete specification for
transport files over ssh, so I'm on early stage.
06 февр. 2015 г. 18:23 пользователь "David Koudela" <
notifications@github.com> написал:

**Comment by bhcleek** [comment]
@vtolstov, there are a couple of options for transporting files over SSH. You can use SCP or SFTP. SCP is essentially RCP, but there's no specification for it. This article, though, is a great primer: https://blogs.oracle.com/janp/entry/how_the_scp_protocol_works

The SFTP specification is documented at https://filezilla-project.org/specs/draft-ietf-secsh-filexfer-02.txt. 

FWIW: A pure Go SFTP client is already implemented: github.com/pkg/sftp.

**Comment by vtolstov** [comment]
Thanks for link i'm try to understand it. I don't think about sftp because
packer for upload uses scp. So i think about scp too.
May be then packer has used sftp internally, i'm write sftp downloader.


2015-02-08 22:30 GMT+03:00 Billie Cleek <notifications@github.com>:

> @vtolstov <https://github.com/vtolstov>, there are a couple of options
> for transporting files over SSH. You can use SCP or SFTP. SCP is
> essentially RCP, but there's no specification for it. This article, though,
> is a great primer:
> https://blogs.oracle.com/janp/entry/how_the_scp_protocol_works
>
> The SFTP specification is documented at
> https://filezilla-project.org/specs/draft-ietf-secsh-filexfer-02.txt.
>
> FWIW: A pure Go SFTP client is already implemented: github.com/pkg/sftp.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mitchellh/packer/issues/1903#issuecomment-73427028>.
>



-- 
Vasiliy Tolstov,
e-mail: v.tolstov@selfip.ru
jabber: vase@selfip.ru

**Comment by 0xBF** [comment]
hi, see #1909, please

**Comment by dkoudela** [comment]
Ok, thanks a lot.
When do you expect it is going to be merged to the master branch?

**Comment by dkoudela** [comment]
I have implemented the filedownload provisioner via command execution:
https://github.com/dkoudela/packer-provisioner-filedownload

**Comment by sethvargo** [comment]
Hi @dkoudela 

It is a bit unclear to me the status of this issue. Where can we help?

**Comment by vtolstov** [comment]
as i see this is needs to be closed, and check #1909
